JavaScript/JavaScriptSDK/SendBuffer.ts

```diff
@@ -0,0 +1,205 @@
+﻿/// <reference path="serializer.ts" />
+/// <reference path="Telemetry/Common/Envelope.ts"/>
+/// <reference path="Telemetry/Common/Base.ts" />
+/// <reference path="Contracts/Generated/ContextTagKeys.ts"/>
+/// <reference path="Context/Application.ts"/>
+/// <reference path="Context/Device.ts"/>
+/// <reference path="Context/Internal.ts"/>
+/// <reference path="Context/Location.ts"/>
+/// <reference path="Context/Operation.ts"/>
+/// <reference path="Context/Sample.ts"/>
+/// <reference path="Context/Session.ts"/>
+/// <reference path="Context/User.ts"/>
+/// <reference path="ajax/ajax.ts"/>
+/// <reference path="DataLossAnalyzer.ts"/>
+
+module Microsoft.ApplicationInsights {
+    "use strict";
+
+    export interface ISendBuffer {
+        /**
+         * Enqueue the payload
+         */
+        enqueue: (payload: string) => void;
+
+        /**
+         * Returns the number of elements in the buffer
+         */
+        count: () => number;
+
+        /**
+         * Clears the buffer
+         */
+        clear: () => void;
+
+        getItems: () => string[];
+
+        /**
+         * Build a batch of all queued elements
+         */
+        batchPayloads: () => string;
+
+        /**
+         * Moves items to the SENT_BUFFER.
+         * The buffer holds items which were sent, but we haven't received any response from the backend for them yet. 
+         */
+        markAsSent: (payload: string[]) => void;
+
+        /**
+         * Removes items from the SENT_BUFFER. Should be called on successful response from the backend. 
+         */
+        clearSent: (payload: string[]) => void;
+    }
+
+    /*
+     * An array based send buffer. 
+     */
+    export class ArraySendBuffer implements ISendBuffer {
+        private _config: ISenderConfig;
+        private _buffer: string[];
+
+        constructor(config: ISenderConfig) {
+            this._config = config;
+
+            this._buffer = [];
+        }
+
+        public enqueue(payload: string) {
+            this._buffer.push(payload);
+        }
+
+        public count(): number {
+            return this._buffer.length;
+        }
+
+        public clear() {
+            this._buffer.length = 0;
+        }
+
+        public getItems(): string[] {
+            return this._buffer;
+        }
+
+        public batchPayloads(): string {
+            if (this.count() > 0) {
+                var batch = this._config.emitLineDelimitedJson() ?
+                    this._buffer.join("\n") :
+                    "[" + this._buffer.join(",") + "]";
+
+                return batch;
+            }
+
+            return null;
+        }
+
+        public markAsSent(payload: string[]) {
+            this.clear();
```

Comment:
I think I can see a semantic readability value to this method but in practice it isn't really doing what the name implies.